### PR TITLE
Refactoring std::chrono out.

### PR DESCRIPTION
**Pull request by nicolasnoble**

### CODE REVIEW

File: test/cpp/end2end/end2end_test.cc [hunk 1]
```diff
@@ -34,10 +34,10 @@
 #include <chrono>
```
Comment (vjpai) [inline test/cpp/end2end/end2end_test.cc]: This one keeps features from chrono?

**Comment by vjpai** [comment]
Could you fill in some info about how a C++ programmer from a non-chrono platform would interface with the API now? It seems that users of such a platform will be directly using gpr_timespec: is this the intention? Do you anticipate some replication or specialization of tests?

**Comment by nicolasnoble** [comment]
Actually, I should probably add a test for this one. But the basic idea is that there's two main locations where we were using std::chrono in our C++ code: AsyncNext and the credential token duration.

The first one is now handled through a template, which can take both std::chrono::system_clock::time_point, and gpr_timespec. So you can call AsyncNext with either one - and a non-chrono supported platform would only have the latter option.

The second one, I changed it for a long. So people can either fill in the actual number of seconds there - like I did in a few tests - or do std::chrono::seconds(3600).count() for example. This last change is really the API-breaking one, and I am doing it for the good reason that in the backend, we're casting an std::chrono::seconds, which is by standard "at least 35 bits", to a long, which can be 32 bits. Meaning that if you're trying to create an infinite token, you're in for a world of hurt in terms of portability, because we're silently downcasting that value.

File: test/cpp/end2end/end2end_test.cc (hunk 1 above)
Comment (nicolasnoble) [inline test/cpp/end2end/end2end_test.cc]: I missed it, thanks :-)

Comment (nicolasnoble) [inline test/cpp/end2end/end2end_test.cc]: Aaand in a few other places in fact. Thanks for pointing it out. The tests currently all don't compile under Windows, so it wasn't obvious this was still breaking.

**Comment by nicolasnoble** [comment]
New code uploaded.

File: include/grpc++/client_context.h [hunk 2]
```diff
@@ -34,15 +34,13 @@
...
 
-#include <chrono>
 #include <map>
 #include <string>
 
 #include <grpc/support/log.h>
 #include <grpc/support/time.h>
 #include <grpc++/config.h>
-
-using std::chrono::system_clock;
```
Comment (ctiller) [inline include/grpc++/client_context.h]: Good.

File: include/grpc++/credentials.h [hunk 3]
```diff
@@ -103,7 +102,7 @@ std::unique_ptr<Credentials> ComputeEngineCredentials();
 // grpc_max_auth_token_lifetime or will be cropped to this value.
 std::unique_ptr<Credentials> ServiceAccountCredentials(
     const grpc::string& json_key, const grpc::string& scope,
-    std::chrono::seconds token_lifetime);
+    long token_lifetime);
```
Comment (ctiller) [inline include/grpc++/credentials.h]: rename argument to token_lifetime_seconds, or otherwise document units

File: include/grpc++/credentials.h [hunk 4]
```diff
@@ -111,7 +110,7 @@ std::unique_ptr<Credentials> ServiceAccountCredentials(
 // this credentials.  It should not exceed grpc_max_auth_token_lifetime or
 // will be cropped to this value.
 std::unique_ptr<Credentials> JWTCredentials(
-    const grpc::string& json_key, std::chrono::seconds token_lifetime);
+    const grpc::string& json_key, long token_lifetime);
```
Comment (ctiller) [inline include/grpc++/credentials.h]: Same here

Comment (nicolasnoble) [inline include/grpc++/credentials.h]: Done.

File: include/grpc++/credentials.h (hunk 3 above)
Comment (nicolasnoble) [inline include/grpc++/credentials.h]: Done.

**Comment by soltanmm** [comment]
cc #1175

File: include/grpc++/time.h [hunk 5]
```diff
@@ -34,6 +34,48 @@
...
 
+#include <grpc++/config.h>
+
+namespace grpc {
+
+/* If you are trying to use CompletionQueue::AsyncNext with a time class that
+   isn't either gpr_timespec or std::chrono::system_clock::time_point, you
+   will most likely be looking at that comment as your compiler will have
+   fired an error below. In order to fix that issue, you have two potential
+   solutions:
```
Comment (vjpai) [inline include/grpc++/time.h]: Minor nit: most of the "that" above should be "this" -- "this comment" ,  "this issue"

File: src/cpp/client/secure_credentials.cc [hunk 6]
```diff
@@ -81,27 +81,27 @@ std::unique_ptr<Credentials> ComputeEngineCredentials() {
 // Builds service account credentials.
 std::unique_ptr<Credentials> ServiceAccountCredentials(
     const grpc::string& json_key, const grpc::string& scope,
-    std::chrono::seconds token_lifetime) {
-  if (token_lifetime.count() <= 0) {
+    long token_lifetime) {
```
Comment (vjpai) [inline src/cpp/client/secure_credentials.cc]: This should probably also be token_lifetime_seconds as in the header

**Comment by vjpai** [comment]
Just wondering if you had had a chance to add in a test for the new platform. That's the only thing remaining here, I think (except for the small nits I mentioned)

**Comment by ctiller** [comment]
What's the status here?

**Comment by vjpai** [comment]
Thanks for the followup; as I mentioned, I was willing to merge last week but was waiting on those few nits and a test for the new case (or at least to be told that it's already tested on some other path). Looks like the code has diverged since then, though, and will end up being harder to merge now.

**Comment by nicolasnoble** [comment]
New code uploaded.

**Comment by nicolasnoble** [comment]
Nobody likes my PRs ? :-P

**Pull request merged**